homeassistant/components/unifiprotect/binary_sensor.py

```diff
@@ -0,0 +1,273 @@
+        ProtectDeviceBinarySensor,
+        camera_descs=CAMERA_SENSORS,
+        light_descs=LIGHT_SENSORS,
+        sense_descs=SENSE_SENSORS,
+    )
+    entities += _async_nvr_entities(data)
+
+    async_add_entities(entities)
+
+
+@callback
+def _async_nvr_entities(
+    data: ProtectData,
+) -> list[ProtectDeviceEntity]:
+    entities: list[ProtectDeviceEntity] = []
+    device = data.api.bootstrap.nvr
+    for index, _ in enumerate(device.system_info.storage.devices):
+        for description in DISK_SENSORS:
+            entities.append(
+                ProtectDiskBinarySensor(data, device, description, index=index)
+            )
+            _LOGGER.debug(
+                "Adding binary sensor entity %s",
+                (description.name or "{index}").format(index=index),
+            )
+
+    return entities
+
+
+class ProtectDeviceBinarySensor(ProtectDeviceEntity, BinarySensorEntity):
+    """A UniFi Protect Device Binary Sensor."""
+
+    def __init__(
+        self,
+        data: ProtectData,
+        description: ProtectBinaryEntityDescription,
+        device: Camera | Light | Sensor | None = None,
+    ) -> None:
+        """Initialize the Binary Sensor."""
+
+        if device and not hasattr(self, "device"):
+            self.device: Camera | Light | Sensor = device
+        self.entity_description: ProtectBinaryEntityDescription = description
+        super().__init__(data)
+        self._doorbell_callback: CALLBACK_TYPE | None = None
+
+    @callback
+    def _async_update_extra_attrs_from_protect(self) -> dict[str, Any]:
+        attrs: dict[str, Any] = {}
+        key = self.entity_description.key
+
+        if key == _KEY_DARK:
+            return attrs
+
+        if key == _KEY_DOORBELL:
+            assert isinstance(self.device, Camera)
+            attrs[ATTR_LAST_TRIP_TIME] = self.device.last_ring
+        elif isinstance(self.device, Sensor):
+            if key in (_KEY_MOTION, _KEY_DOOR):
+                if key == _KEY_MOTION:
+                    last_trip = self.device.motion_detected_at
+                else:
+                    last_trip = self.device.open_status_changed_at
+
+                attrs[ATTR_LAST_TRIP_TIME] = last_trip
+        elif isinstance(self.device, Light):
+            if key == _KEY_MOTION:
+                attrs[ATTR_LAST_TRIP_TIME] = self.device.last_motion
+
+        return attrs
+
+    @callback
+    def _async_update_device_from_protect(self) -> None:
+        super()._async_update_device_from_protect()
+
+        assert self.entity_description.ufp_value is not None
+
+        self._attr_extra_state_attributes = (
+            self._async_update_extra_attrs_from_protect()
+        )
+
+        if self.entity_description.key == _KEY_DOORBELL:
+            last_ring = get_nested_attr(self.device, self.entity_description.ufp_value)
+            now = utcnow()
+
+            is_ringing = (
+                False if last_ring is None else (now - last_ring) < RING_INTERVAL
```

Comment:
As I said in the earlier review, we don't allow calculating state in device and service interfacing integrations. I'm surprised we merged this PR with this included.

Reply:
Based on conversation in the Discord, I think it does make more sense to turn this in to a timestamp sensor, but I would rather hold off on doing that until 2022.3 if you are okay with that. That way there are not breaking changes from HACS to the HA Core version. To increase adoption. A lot of users relay on this sensor, it is probably the most popular one aside from the motion ones (TBA).

Reply:
No, we shouldn't include features that are not in line with our architecture design standards.

This is unfortunately one of the problems with publishing custom integrations and later moving them to core.

Reply:
We got too caught up in the how its implemented rather if it should be in the first place 🙈 

@AngellusMortis  Please reach out on discord to discuss. There are a few options:

- We could change this to a device trigger, unfortunately this is not a good UX yet since we don't have a way to classify them yet so they can't get linked automatically to the camera
- Switch to the timestamp
- Get the end of ring time from the websocket data so the package code is doing the callback instead of the derived version in hass
- probably something else I haven't thought of.